src/controllers/rendering/controllerrenderingengine.cpp

```diff
@@ -0,0 +1,354 @@
+#include "controllers/scripting/legacy/controllerscriptinterfacelegacy.h"
+#include "moc_controllerrenderingengine.cpp"
+#include "qml/qmlwaveformoverview.h"
+#include "util/cmdlineargs.h"
+#include "util/logger.h"
+#include "util/time.h"
+
+namespace {
+const mixxx::Logger kLogger("ControllerRenderingEngine");
+} // anonymous namespace
+
+QMutex ControllerRenderingEngine::s_glMutex = QMutex();
+
+ControllerRenderingEngine::ControllerRenderingEngine(
+        const LegacyControllerMapping::ScreenInfo& info,
+        ControllerScriptEngineBase* parent)
+        : QObject(),
+          m_screenInfo(info),
+          m_GLDataFormat(GL_RGBA),
+          m_GLDataType(GL_UNSIGNED_BYTE),
+          m_isValid(true),
+          m_pControllerEngine(parent) {
+    switch (m_screenInfo.pixelFormat) {
+    case QImage::Format_RGB16:
+        m_GLDataFormat = GL_RGB;
+        m_GLDataType = m_screenInfo.reversedColor
+                ? GL_UNSIGNED_SHORT_5_6_5_REV
+                : GL_UNSIGNED_SHORT_5_6_5;
+        break;
+    case QImage::Format_RGB888:
+        m_GLDataFormat = m_screenInfo.reversedColor ? GL_BGR : GL_RGB;
+        m_GLDataType = GL_UNSIGNED_BYTE;
+        break;
+    case QImage::Format_RGBA8888:
+        m_GLDataFormat = m_screenInfo.reversedColor ? GL_BGRA : GL_RGBA;
+        m_GLDataType = GL_UNSIGNED_BYTE;
+        break;
+    default:
+        m_isValid = false;
+        DEBUG_ASSERT(!"Unsupported format");
+    }
+
+    if (!m_isValid) {
+        return;
+    }
+
+    prepare();
+}
+
+void ControllerRenderingEngine::prepare() {
+    m_pScheduler = std::make_unique<ControllerRenderingScheduler>(m_screenInfo);
+
+    connect(this,
+            &ControllerRenderingEngine::setupRequested,
+            this,
+            &ControllerRenderingEngine::setup);
+    connect(this,
+            &ControllerRenderingEngine::stopRequested,
+            this,
+            &ControllerRenderingEngine::finish);
+    connect(m_pScheduler.get(),
+            &ControllerRenderingScheduler::frameRequested,
+            this,
+            &ControllerRenderingEngine::renderFrame);
+
+    moveToThread(qApp->thread());
+}
+
+ControllerRenderingEngine::~ControllerRenderingEngine() {
+    VERIFY_OR_DEBUG_ASSERT(!m_fbo) {
+        kLogger.warning() << "The ControllerEngine is being deleted but hasn't been "
+                             "cleaned up. Brace for impact";
+    };
+}
+
+void ControllerRenderingEngine::start() {
+    VERIFY_OR_DEBUG_ASSERT(!m_pScheduler->isFinished() &&
+            !m_pScheduler->isInterruptionRequested()) {
+        kLogger.warning() << "Render thread has or ir about to terminate. Cannot "
+                             "start this render anymore.";
+        return;
+    }
+    m_pScheduler->start(QThread::NormalPriority);
+}
+bool ControllerRenderingEngine::isRunning() const {
+    return m_pScheduler && m_pScheduler->isRunning();
+}
+
+void ControllerRenderingEngine::requestSetup(std::shared_ptr<QQmlEngine> qmlEngine) {
```

Comment:
`qmlEngine` could be `nullptr`. especially now that this is called in combination with a dynamic_cast.

Reply:
I'll add a `VERIFY_OR_DEBUG_ASSERT`